spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java

```diff
@@ -0,0 +1,146 @@
+
+	private static final Map<String, Function<Snapshot, Number>> TIMER_SNAPSHOT_METRICS;
+
+	private static final Map<String, Function<Timer, Number>> TIMER_METRICS;
+
+	static {
+		Map<String, Function<Snapshot, Number>> map = new HashMap<>(10);
+		map.put("p999", Snapshot::get999thPercentile);
+		map.put("p99", Snapshot::get99thPercentile);
+		map.put("p98", Snapshot::get98thPercentile);
+		map.put("p95", Snapshot::get95thPercentile);
+		map.put("p75", Snapshot::get75thPercentile);
+		map.put("min", Snapshot::getMin);
+		map.put("max", Snapshot::getMax);
+		map.put("mean", Snapshot::getMean);
+		map.put("median", Snapshot::getMedian);
+		map.put("stdDev", Snapshot::getStdDev);
+		TIMER_SNAPSHOT_METRICS = Collections.unmodifiableMap(map);
+	}
+
+	static {
+		Map<String, Function<Timer, Number>> map = new HashMap<>(4);
+		map.put("mean_rate", Timer::getMeanRate);
+		map.put("m1_rate", Timer::getOneMinuteRate);
+		map.put("m5_rate", Timer::getFiveMinuteRate);
+		map.put("m15_rate", Timer::getFifteenMinuteRate);
+		TIMER_METRICS = Collections.unmodifiableMap(map);
+	}
+
+	private final Iterable<Tag> tags;
+
+	private final CqlSession session;
+
+	public CassandraMetricsBinder(@NonNull CqlSession cqlSession) {
+		this.session = cqlSession;
+		this.tags = Tags.of("sessionName", cqlSession.getName());
+	}
+
+	@Override
+	public void bindTo(@NonNull MeterRegistry meterRegistry) {
+		Assert.isTrue(this.session.getMetrics().isPresent(), "Metrics on the CqlSession needs to be present");
+
+		this.session.getMetrics().get().getRegistry().getMetrics().forEach((name, m) -> {
+			if (m instanceof Timer) {
+				registerTimer(name, (Timer) m, meterRegistry);
+			}
+			else if (m instanceof Counting) {
+				registerMeter(name, (Counting) m, meterRegistry);
+			}
+			else if (m instanceof com.codahale.metrics.Gauge) {
+				registerGauge(name, (com.codahale.metrics.Gauge<?>) m, meterRegistry);
+			}
+		});
+	}
+
+	private void registerTimer(String metricName, Timer timer, MeterRegistry meterRegistry) {
+		String name = String.format("%s.%s", SESSION_PREFIX, metricName);
+
+		// count metric should be registered at the root level of the metric name
+		FunctionCounter.builder(name, timer, Timer::getCount).tags(this.tags).description(DESCRIPTION)
+				.register(meterRegistry);
+
+		TIMER_METRICS.forEach((n, supplier) -> {
+			String nameTimerMetric = String.format("%s.%s", name, n);
+			FunctionCounter.builder(nameTimerMetric, timer, (t) -> supplier.apply(timer).doubleValue()).tags(this.tags)
+					// all rates are calculated per second
+					.baseUnit(TimeUnit.SECONDS.name()).description(DESCRIPTION).register(meterRegistry);
+		});
+
+		final Snapshot snapshot = timer.getSnapshot();
+		TIMER_SNAPSHOT_METRICS.forEach((n, supplier) -> {
+			String nameTimerMetric = String.format("%s.%s", name, n);
+			FunctionCounter.builder(nameTimerMetric, timer, (t) -> supplier.apply(snapshot).longValue()).tags(this.tags)
```

Comment:
Again, not monotonically increasing (except for max) so should not be a FunctionCounter.